bin/add_hypothetical_protein_descriptions.py

```diff
@@ -0,0 +1,714 @@
+        )
+        if func_description:
+            return func_description, source
+    if acc in ipr_memberdb_only and any(
+        db in {"SUPERFAMILY", "Gene3D"}
+        for db in ipr_memberdb_only[acc]["no_type"].keys()
+    ):
+        selected_keys = ["SUPERFAMILY", "Gene3D"]
+        subset_dict = dict()
+        for key in selected_keys:
+            if key in ipr_memberdb_only[acc]["no_type"]:
+                subset_dict[key] = ipr_memberdb_only[acc]["no_type"][key]
+        func_description, source = get_superfamily_info(subset_dict)
+        if func_description:
+            return func_description, source
+    if acc in eggnog_annot:
+        return eggnog_annot[acc], "eggNOG"
+    return "hypothetical protein", "Prokka"
+
+
+def get_description_and_source(my_dict, ipr_type):
+    if "NCBIfam" in my_dict[ipr_type]:
+        func_description, source = pull_out_description(
+            my_dict[ipr_type]["NCBIfam"], "sig_desc", "ipr_desc"
+        )
+        source_description = format_source("NCBIfam", source)
+        return func_description, source_description
+    elif any(key.lower() not in {"superfamily", "gene3d"} for key in my_dict[ipr_type]):
+        highest_match = get_best_match(my_dict[ipr_type])
+        db = next(iter(highest_match))
+        func_description, source = pull_out_description(
+            highest_match[db], "ipr_desc", "sig_desc"
+        )
+        source_description = format_source(db, source)
+        return func_description, source_description
+    else:
+        return "", ""
+
+
+def get_superfamily_info(my_dict):
+    if len(my_dict.keys()) == 1:
+        db = next(iter(my_dict.keys()))
+        description, source = pull_out_description(my_dict[db], "ipr_desc", "sig_desc")
+        source_description = format_source(db, source)
+    else:
+        db = find_higher_match(my_dict)
+        description, source = pull_out_description(my_dict[db], "ipr_desc", "sig_desc")
+        source_description = format_source(db, source)
+    return description, source_description
+
+
+def format_source(db, source):
+    if source == "ipr_desc":
+        return "InterPro({})".format(db)
+    else:
+        return db
+
+
+def get_best_match(ipr_dict):
+    best_fraction = 0
+    best_level = 0
+    highest_match = dict()
+    for db in ipr_dict:
+        if ipr_dict[db]["level"] is None:
+            ipr_dict[db]["level"] = 0
+        if not db.lower() in ["superfamily", "gene3d"] and (
+            ipr_dict[db]["sig_desc"] != "-" or ipr_dict[db]["ipr_desc"] != "-"
+        ):
+            if ipr_dict[db]["level"] > best_level or (
+                ipr_dict[db]["level"] == best_level
+                and ipr_dict[db]["match"] > best_fraction
+            ):
+                best_fraction = ipr_dict[db]["match"]
+                best_level = ipr_dict[db]["level"]
+                highest_match = dict()
+                highest_match[db] = ipr_dict[db]
+    if "Pfam" not in highest_match and best_fraction > 0.30:
```

Comment:
I would put the thresholds as constants at the top of the file, making it easier to stop and modify if needed.